### DISCUSSION TITLE
Should clangd's tests be able to make use of the standard library?

### DISCUSSION THREAD

**Discussion by HighCommander4** [clangd/clangd#940] [General]
I've noticed that several new C++ language features require some library support to work. A well-known case of this is list-initialization requiring `std::initializer_list`, but some newer ones include three-way comparisons (`<=>`) requiring `std::strong_equality` and similar types, and C++20 coroutines requiring types to specialize `std::coroutine_traits` in order for e.g. a `co_await` expression to compile.

When writing tests for such language features in clangd, our approach so far has been to copy a minimal version of the required library code into the test, but this gets cumbersome as more features require such code and the required code gets more complex.

Given that LLVM includes a standard library implementation, is there a way to make it visible to our tests, such that tests for relevant language features can just include the relevant header (e.g. `#include <coroutine>`) and keep the test code itself focused on the language feature being tested?

**Comment by sam-mccall** [comment]
The historical reason is definitely just it's the simplest thing, and I agree that with coroutines and others it's no longer simple.

That said I'm not sure real libc++ is the best approach for tests. When debugging clangd issues, examples involving the standard library usually don't count as "reduced" - to get at the essence of it, I often end up bisecting/hacking up the library code as if it were any other library. The metaprogramming and obfuscation in the standard library is not really something I'd like to consider when looking at a test failure.

clang-tidy has reusable "mock" headers for this purpose, that provide simple good-enough approximations of critical parts of the standard library. Only the needed symbols, no clever metaprogramming or macros. What do you think of us doing the same for these headers? We could wire up TestTU so they'd just be available as `#include <coroutine>` with no per-test setup.

In clang-tidy there have been instances there of bugs that were obscured because the mocks are too simple, but we can't entirely solve that: e.g. using libc++ will still leave us open to unexpected libstdc++ interactions.

(If we *were* to depend on the real libc++, we could consider bundling it for indexing/fallback use when there's no stdlib available/...)

**Reply by HighCommander4** [reply]
Thanks, mock headers sound promising and they have the notable advantage that they're simpler if we need to e.g. debug the instantiation of a library template.
